[google_maps_flutter] Add ability to animate camera with duration

jokerttu opened the pull request:
Adds ability to configure camera animation duration on Android and iOS.

Resolves [#39810](https://github.com/flutter/flutter/issues/39810)
Resolves [#44284](https://github.com/flutter/flutter/issues/44284)

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I [linked to at least one issue that this PR fixes] in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style], or this PR is [exempt from CHANGELOG changes].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[Discord]: https://github.com/flutter/flutter/blob/master/docs/contributing/Chat.md
[linked to at least one issue that this PR fixes]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#overview
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#version
[following repository CHANGELOG style]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#changelog-style
[exempt from CHANGELOG changes]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#changelog
[test-exempt]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#tests

jokerttu commented:
Web support is not included, as there are still some areas open for discussion.
Test implementation here:
https://github.com/CodemateLtd/packages/compare/feature/animate-camera-duration...CodemateLtd:packages:feature/animate-camera-duration-web?expand=1

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/method_channel/method_channel_google_maps_flutter.dart:
```diff
@@ -421,8 +421,20 @@ class MethodChannelGoogleMapsFlutter extends GoogleMapsFlutterPlatform {
   }) {
-    return channel(mapId).invokeMethod<void>('camera#animate', <String, Object>{
+    return channel(mapId).invokeMethod<void>('camera#animate',
+        <String, dynamic>{'cameraUpdate': cameraUpdate.toJson()});
```
Why was this changed to `dynamic`? We only use `dynamic` when we have an extremely compelling reason to do so as it bypasses a lot of static analysis.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/method_channel/method_channel_google_maps_flutter.dart:
```diff
@@ -421,8 +421,20 @@ class MethodChannelGoogleMapsFlutter extends GoogleMapsFlutterPlatform {
+    return channel(mapId)
+        .invokeMethod<void>('camera#animate', <String, dynamic>{
       'cameraUpdate': cameraUpdate.toJson(),
+      'duration': configuration.duration?.inMilliseconds,
```
We don't generally want to send nulls in collections in method channels.

I would recommend just reverting the changes to this file; we don't use the method channel implementation, it's only still here because removing it requires a breaking change. There's no obligation to implement new features for it.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/platform_interface/google_maps_flutter_platform.dart:
```diff
@@ -194,6 +194,20 @@ abstract class GoogleMapsFlutterPlatform extends PlatformInterface {
+  /// platform side.
+  Future<void> animateCameraWithConfiguration(
+    CameraUpdate cameraUpdate, {
+    required CameraUpdateAnimationConfiguration configuration,
```
This can be positional rather than named since it's a core parameter.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/platform_interface/google_maps_flutter_platform.dart:
```diff
@@ -194,6 +194,20 @@ abstract class GoogleMapsFlutterPlatform extends PlatformInterface {
+    required int mapId,
+  }) {
+    throw UnimplementedError(
+        'animateCameraWithConfiguration() has not been implemented.');
```
This should call `animateCamera`; see https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#changing-platform-interface-method-parameters

stuartmorgan replied:
And the forwarding should get a unit test in this package.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter/lib/google_maps_flutter.dart:
```diff
@@ -24,6 +24,7 @@ export 'package:google_maps_flutter_platform_interface/google_maps_flutter_platf
         CameraPositionCallback,
         CameraTargetBounds,
         CameraUpdate,
+        CameraUpdateAnimationConfiguration,
```
We generally don't want to expose these parameter objects at the app-facing package API level; they solve a problem (adding optional parameter being a breaking change for subclasses) that doesn't exist at this level, and add complexity for all clients (vs. just ourselves in implementation packages). The client-facing API should expose the duration directly, and that should be packed into the configuration object as an internal implementation detail.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter/example/lib/animate_camera.dart:
```diff
@@ -156,14 +180,33 @@ class AnimateCameraState extends State<AnimateCamera> {
+            ),
+            const SizedBox(width: 5),
+            Switch(
+              value: _cameraUpdateAnimationConfiguration != null,
```
These should all be "Duration" rather than "Configuration" once the other change is made.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter/example/lib/animate_camera.dart:
```diff
@@ -156,14 +180,33 @@ class AnimateCameraState extends State<AnimateCamera> {
+          mainAxisAlignment: MainAxisAlignment.center,
+          children: <Widget>[
+            const Text(
+              'With configuration\n(10 second duration)',
```
`'With 10 second duration'`

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter/example/integration_test/src/maps_inspector.dart:
```diff
@@ -612,6 +627,186 @@ void runTests() {
+      await tester.pumpAndSettle();
+    },
+    variant: _cameraUpdateTypeVariants,
+    skip: kIsWeb,
```
This will need a tracking issue. Or is the plan to have web support ready before the app-facing PR lands?

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter/example/integration_test/src/maps_inspector.dart:
```diff
@@ -612,6 +627,186 @@ void runTests() {
+          StreamQueue<void>(cameraFiredStream.stream);
+
+      // We set the duration to 100ms and check that the animation is completed
+      // faster than the normal animation (~300ms on Android) to make sure the
```
This is run on other platforms as well, and it's not clear from the comment why this test is correct on those platforms.

jokerttu replied:
Comment updated

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter/example/integration_test/src/maps_inspector.dart:
```diff
@@ -612,6 +627,186 @@ void runTests() {
+      const int cameraAnimationDurationMS = 100;
+
+      // The threshold is set to 200ms because the normal camera animation should
+      // take more than 200ms on android.
```
This is very likely to become a flaky test as written, since it assumes things about the timing of things beyond the animation itself (multiple async steps with cross-thread hops, potentially delays within the Google Maps SDK itself), and the tolerance here is relatively tight for environments like an emulator. It's also not ideal (per the previous comment) to have a test that would become subject to false negatives if a value we don't control in a different SDK changed.

What about having the test run once with an extremely short time, like 10ms, and once with a much longer time like a second, and make sure the difference between them is more than, say, half the larger value. That gives a fair amount of tolerance, but should still easily catch the value being ignored.

jokerttu replied:
I used suggested 10ms and 1000ms durations and tests seems to pass.
But the duration increase caused CI tests to timeout
@stuartmorgan, how to proceed with this?

stuartmorgan replied:
Increasing a few integration test durations by a second shouldn't have caused timeouts; are you sure there wasn't some other change that's causing timeouts now?

jokerttu replied:
@stuartmorgan 
Is there a way to access more detailed test output than this: https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8729373696952662689/+/u/Run_package_tests/Firebase_Test_Lab/stdout
(I don't have access to the firebase console: https://console.firebase.google.com/project/flutter-infra-staging/testlab/histories/bh.8eb89ae93eb3eb36/matrices/6375674335592449274)

Tests run fine locally without timeouts.

Running tests for all CameraUpdateType variants (9 types) adds at least 12 seconds per package due to the testing with 200ms and 1000ms animation delays + map initialization. 

Should we consider testing just one type here since implementations between CameraUpdate types are identical on iOS and Android?
On web implementation testing each variation is more important because there is separate method for each camera update type.

stuartmorgan replied:
Unfortunately the console isn't showing anything interesting that I can see. The way integration tests run in FTL we don't get incremental test output; so it's mostly device-level logs. For the rare cases like this where an issue happens in FTL but not locally, sometimes we have to do guess-and-check debugging unfortunately, by pushing alternate versions (I would suggest a new PR so you don't pollute this one with unwanted changes) with changes. E.g., try just a minimal new test, see if it works, then add more incrementally. It's not ideal, but the FTL-style run just doesn't always give us the info we need.

You could also try adding some native logging so that we can look for that in the device logs, to get a sense of what is happening on the native side. E.g., maybe log before each animated move request, with some details about the args to try to identify it, and also before sending each camera idle notification to Dart.

> Running tests for all CameraUpdateType variants (9 types) adds at least 12 seconds per package due to the testing with 200ms and 1000ms animation delays + map initialization.
> 
> Should we consider testing just one type here since implementations between CameraUpdate types are identical on iOS and Android?

We could, but for context the end-to-end integration test run for `google_maps_flutter_android` right now in CI is about 6 minutes, so 12 seconds of additional tests aren't really that big a deal.

jokerttu replied:
Tested locally 
- with one CameraUpdate variant, all tests runned in average 48 seconds. (tested 3 times)
- with all 9 CameraUpdate variants, all tests runned in average 1 minute 22 seconds (tested 3 times)

I bet there is 1 minute timeout set for the tests.

Should we remove variants and test only with one CameraUpdate type?
@stuartmorgan ?

stuartmorgan replied:
I was curious, so I just pulled this branch and tried it locally, and see it hang in my simulator. I'll debug a bit and see what I find.

stuartmorgan replied:
What I see locally is that `await cameraFiredQueue.next;` never completes.

Are you testing with a maps key? We don't currently test with one in CI because in practice everything we were testing worked without it, but maybe this is something that doesn't work without a key?

jokerttu replied:
@stuartmorgan 
`animateCamera` does not move the camera at all on Android if Google Maps API key is not given. 

Should we skip these tests on android for now?

stuartmorgan replied:
In theory yes (https://github.com/flutter/flutter/issues/131071), but it will require some coordination with infra-level changes so it's not something I expect to happen in the very short term. We could try to prioritize it in Q1 of next year, potentially.

For now it's fine to skip this test on Android with a reference to that issue in a comment.

jokerttu replied:
Skipping tests on Android
Tests works fine with api key, when enabled.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter/example/integration_test/src/maps_inspector.dart:
```diff
@@ -636,3 +831,94 @@ Marker _copyMarkerWithClusterManagerId(
 }
+
+CameraUpdate _getCameraUpdateForType(CameraUpdateType type) {
+  switch (type) {
```
Consider using the newer `return switch (type) {...}` pattern syntax.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter/example/integration_test/src/maps_inspector.dart:
```diff
@@ -636,3 +831,94 @@ Marker _copyMarkerWithClusterManagerId(
+  GoogleMapController controller,
+  Matcher Function(Matcher matcher) wrapMatcher,
+) async {
+  // The target might differ a bit from the expected target; this is why we use
```
"; this is why we" -> ", so"

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_android/pigeons/messages.dart:
```diff
@@ -103,6 +103,13 @@ class PlatformCameraUpdateZoomTo {
 }
 
+/// Pigeon representation of a CameraUpdateAnimationConfiguration.
+class PlatformCameraUpdateAnimationConfiguration {
```
Pigeon interfaces are internal APIs, and can be changed at will, so there is no need to use parameter objects in general; the method can just take a durationMilliseconds parameter directly (same on iOS).

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/example/ios14/ios/RunnerTests/GoogleMapsTests.m:
```diff
@@ -86,6 +87,122 @@ - (void)testHandleResultTileDownsamplesWideGamutImages {
+  // Assert that the camera zoom is updated after the animation completes
+  XCTAssertEqual(controller.mapView.camera.zoom, 10.0,
+                 "Camera zoom should be updated to 10.0 after animation.");
+}
```
Actually running animations is pretty slow for unit tests; could we just dependency inject a simple wrapper around `CATransaction` and ensure that's being called as expected, rather than testing the animation itself (which is largely testing the underlying SDK rather than the plugin code)?

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/pigeons/messages.dart:
```diff
@@ -51,6 +51,13 @@ class PlatformCameraUpdate {
 }
 
+/// Pigeon representation of a CameraUpdateAnimationConfiguration.
+class PlatformCameraUpdateAnimationConfiguration {
```
Same note here as on Android.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_web/lib/src/google_maps_flutter_web.dart:
```diff
@@ -132,6 +132,15 @@ class GoogleMapsPlugin extends GoogleMapsFlutterPlatform {
+    required CameraUpdateAnimationConfiguration configuration,
+    required int mapId,
+  }) {
+    return animateCamera(cameraUpdate, mapId: mapId);
```
This will need a TODO referencing an issue, as with the test disabling.

ditman replied:
Unfortunately, I don't think Google Maps on the web supports *any* control of the duration of the animation... See docs for `moveCamera`, `panTo` and friends:

* https://developers.google.com/maps/documentation/javascript/reference/map#Map.moveCamera

So I think you can remove the TODO, the link and everything. Maybe we can log that the value will be silently ignored to the js debug console to let users know? Or at least document in the method of the plugin that the value will not be used on the web.

jokerttu replied:
Discussion about this on web-related issue here: https://github.com/flutter/flutter/issues/159265#issuecomment-2580116988

jokerttu replied:
Web implementation is not necessary as platform interface now automatically defaults to 
`animateCamera`
https://github.com/flutter/packages/pull/7648/files#diff-82107f54c1af7174644137b415b811aa06290c54af28a3a81de0589ca9915e63R207
I will remove web-implementation from this PR as there is separate issue for web impementation: https://github.com/flutter/flutter/issues/159265

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/GoogleMapController.java:
```diff
@@ -937,12 +937,21 @@ public void moveCamera(@NonNull Messages.PlatformCameraUpdate cameraUpdate) {
+          null);
+    } else {
+      googleMap.animateCamera(Convert.cameraUpdateFromPigeon(cameraUpdate, density));
+    }
```
There should be a native unit test validating the branching here (via a mock map).

camsim99 commented on packages/google_maps_flutter/google_maps_flutter_android/example/integration_test/google_maps_tests.dart:
```diff
@@ -1446,6 +1463,221 @@ void googleMapsTests() {
+      );
+
+      // Check that position is not updated immediately to the target.
+      final CameraPosition beforeFinishedPosition =
```
I have the same question here about flakiness as the test before. Is that concern why this is only checked for the longer animation and not the shorter?

jokerttu replied:
Short period is so short, that the might be possilbity that the latencies on communication and SDK might delay the camera position query so much that it is already animated to target location, causing flaky tests. 
For longer duration we could safely say that communication latencies are not causing 1 second delays.
Camera animation is not affected by networking either.

camsim99 commented on packages/google_maps_flutter/google_maps_flutter_android/example/integration_test/google_maps_tests.dart:
```diff
@@ -1446,6 +1463,221 @@ void googleMapsTests() {
+      await cameraFiredQueue.next;
+      expect(cameraIdleFired, isTrue);
+
+      // For short animation duration, check that the animation is completed
+      // shower than the middle point of the animation test durations.
```
I think this was meant to be:
```suggestion
      // For longer animation duration, check that the animation is completed
      // slower than the middle point of the animation test durations.
```

camsim99 commented on packages/google_maps_flutter/google_maps_flutter_android/example/integration_test/google_maps_tests.dart:
```diff
@@ -1446,6 +1463,221 @@ void googleMapsTests() {
+      expect(stopwatch.elapsedMilliseconds,
+          greaterThan(animationDurationMiddlePoint));
+
+      // Check that position is animated after the animation is done.
```
Nit:
```suggestion
      // Check that position is updated after the animation is done.
```

camsim99 reviewed (commented):
The changes LGTM :) but left some questions/comments on the stability of the integration test.

jokerttu commented on packages/google_maps_flutter/google_maps_flutter_android/example/integration_test/google_maps_tests.dart:
```diff
@@ -1446,6 +1463,221 @@ void googleMapsTests() {
+        _getCameraUpdateForType(_cameraUpdateTypeVariants.currentValue!);
+    await controller.animateCamera(cameraUpdate);
+
+    // Check that position is not updated immediately to the target.
```
Improved tests documentation to explain better why this is done.
And yes this could possible be flaky, but most likely not.
On android the default animation time is about 300ms. This should be long enought time to mitigate possible communication or SDK delays.
Reason why this is tested, is to mitigate out possiblity that camera is not animated, and is just moved instantly to new location. We could have also used stopwatch and wait cameraIdleEvent, and see if the event does not come too soon, but then we would need to hardcode limits.
It this test is removed, we can only test that camera is moved to right position, but cannot be aware if it is animated or not.

jokerttu commented on packages/google_maps_flutter/google_maps_flutter_android/example/integration_test/google_maps_tests.dart:
```diff
@@ -1446,6 +1463,221 @@ void googleMapsTests() {
+      // For short animation duration, check that the animation is completed
+      // faster than the middle point of the animation test durations.
+      expect(stopwatch.elapsedMilliseconds,
+          lessThan(animationDurationMiddlePoint));
```
This solution is based on suggestion by @stuartmorgan https://github.com/flutter/packages/pull/7648#discussion_r1833390127
Testing with two times mitigate away the possiblity that animateCamera by default animates with same speed.
The duration won't be exactly same as detected by stopwatch, because internal delays on messaging and SDK functionality, so picking proper threshold for testing duration could not work.

Most likely having one test run with longer period, lets say 1 second, could work, but this assumes that default animation speed on each platform is different.

jokerttu commented:
Updated integration tests, by using Completer instead of StreamQueue to detect OnCameraIdle events.
This improved the test quality, and removed need to drain queue, which was possibly a bit flaky solution.

Integration tests are also now skipped on Android as they work only with Maps API Key which is not available on CI at the moment.
Tests passes on both projects for Android if enabled and Maps API Key is available.

camsim99 reviewed (commented):
Android changes LGTM! Thanks @jokerttu for the thorough explanations :)

devnoaman commented:
any update ?

jokerttu commented:
This is still awaiting iOS review from @cbracken (code owner) before I can proceed with splitting this into sub-PRs.

ditman commented on packages/google_maps_flutter/google_maps_flutter_web/lib/src/google_maps_flutter_web.dart:
```diff
@@ -362,6 +373,7 @@ class GoogleMapsPlugin extends GoogleMapsFlutterPlatform {
     GoogleMapsInspectorPlatform.instance = GoogleMapsInspectorWeb(
       (int mapId) => _map(mapId).configuration,
       (int mapId) => _map(mapId).clusterManagersController,
+      (int mapId) => _map(mapId),
```
Now that we're passing the whole map controller, the constructor for the `GoogleMapsInspectorWeb` could be simplified to just accept the controller, and have it internally extract `.configuration` and `.clusterManagersController`, as needed, I think.

ditman replied:
I think this is fine, that way we don't accumulate too much technical debt from the web implementation, or have too many `skip: kIsWeb` tests in the app-facing package! Thanks!

ditman reviewed (commented):
Web bits look OK to me, I had only a couple of minor remarks. Thanks!

reidbaker reviewed (commented):
Android code looks good. Deferring android final approval to @camsim99

jokerttu commented on packages/google_maps_flutter/google_maps_flutter_android/android/src/test/java/io/flutter/plugins/googlemaps/GoogleMapControllerTest.java:
```diff
@@ -248,4 +255,47 @@ public void UpdateHeatmaps() {
+    }
+
+    verify(mockGoogleMap, times(1))
+        .animateCamera(any(CameraUpdate.class), eq(durationMilliseconds.intValue()), isNull());
```
animateCamera of googleMap instance is called like this:
```java
      googleMap.animateCamera(
          Convert.cameraUpdateFromPigeon(cameraUpdate, density),
          durationMilliseconds.intValue(),
          null);
```

This line only verifies that this method is called as supposed.
The last item is cancellable callback that is not currently supported:
https://developers.google.com/android/reference/com/google/android/gms/maps/GoogleMap#public-final-void-animatecamera-cameraupdate-update,-int-durationms,-googlemap.cancelablecallback-callback

jokerttu commented on packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/platform_interface/google_maps_inspector_platform.dart:
```diff
@@ -137,4 +137,13 @@ abstract class GoogleMapsInspectorPlatform extends PlatformInterface {
     throw UnimplementedError('getClusters() has not been implemented.');
   }
+
+  /// Returns current camera position.
```
@ditman should I add following method here and use it in tests to see if getting camera position is supported on the platform.
Then I can drop the changes for web platform from this PR and everything on it's separate PR later after problems with it are solved?

```dart
  /// If the platform supports getting camera position
  bool supportsGettingGameraPosition() => false;
```

stuartmorgan replied:
That sounds good; pretty much any time we can isolate platform-specific limitations behind a generic support query it makes the rest of the code more maintainable.

jokerttu replied:
Done at https://github.com/flutter/packages/pull/7648/commits/f4a39ce081257bcbfe2d38e9ffa637023c2f318e

jokerttu commented:
As web platform changes are now dropped, this still needs review/approval from
Android: @reidbaker https://github.com/flutter/packages/blob/main/CODEOWNERS#L78
iOS: @cbracken https://github.com/flutter/packages/blob/main/CODEOWNERS#L95
App facing package: @stuartmorgan

camsim99 reviewed (approved):
Sorry I just commented before. Android LGTM! :)

stuartmorgan reviewed (commented):
I went ahead and did the iOS review, and it LGTM; you should be good to split out sub PRs now.

jokerttu commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/GoogleMapController_Test.h:
```diff
@@ -4,9 +4,17 @@
 
 NS_ASSUME_NONNULL_BEGIN
 
+@interface FGMMapCallHandler : NSObject <FGMMapsApi>
```
@stuartmorgan please check following commit if this addresses this issue.
https://github.com/flutter/packages/pull/7648/commits/d235debbf2bbdac667a3571e5db561f882771587

Commit exposes `FGMMapCallHandler` in `GoogleMapController.h` that allows using `(Test)` extension in `GoogleMapController_Test.h`.

stuartmorgan replied:
Yes, thanks.

jokerttu commented:
There is now splitted PR for platform interface implementation at https://github.com/flutter/packages/pull/8596

stuartmorgan reviewed (approved):
LGTM!
